File: plugins/cordova-plugin-muzik/src/android/AutoPlayBridge.java
package com.muzik.androidplugin;

import com.muzik.accessory.MzAccessory;
import com.muzik.accessory.callback.IAutoPlayCallback;

import java.util.LinkedList;
import java.util.Queue;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;

public class AutoPlayBridge implements IAutoPlayCallback {

    private MzAccessory mzAccessory;
    private Queue<CallbackContext> callbacks;

    public AutoPlayBridge(MzAccessory mzAccessory) {
        this.mzAccessory = mzAccessory;
        this.callbacks = new LinkedList<CallbackContext>();
    }

    public synchronized void getAutoPlaySetting(CallbackContext callback) {
        this.callbacks.add(callback);
        this.mzAccessory.getAutoPlaySetting(this);
    }

    public synchronized void setAutoPlaySetting(CallbackContext callback, boolean autoPlay) {
        this.callbacks.add(callback);
        this.mzAccessory.setAutoPlaySetting(this, autoPlay);
    }
    
    @Override
    public synchronized void onResponseReceived(boolean isAutoPlay) {
        if (!this.callbacks.isEmpty()) {
            PluginResult result = new PluginResult(PluginResult.Status.OK, isAutoPlay);
            this.callbacks.poll().sendPluginResult(result);
        }
    }
}

File: plugins/cordova-plugin-muzik/src/ios/src/AutoPlayBridge.m
//
//  AutoPlayBridge.g
//

#import "AutoPlayBridge.h"
#import "Queue.h"

@interface AutoPlayBridge ()

@property (nonatomic, strong) MZAccessory* mzAccessory;
@property (nonatomic, strong) CDVPlugin* plugin;
@property (nonatomic, strong) Queue* commandQueue;

@end

@implementation AutoPlayBridge

// Default constructor
- (id) init {
    self = [super init];
    if (self) {
        self.mzAccessory = nil;
        self.plugin = nil;
        self.commandQueue = [[Queue alloc] init];
    }
    return self;
}

// Constructor with references to an instance of MZAccessory, and CDVPlugin
- (id) initWithAccessory:(MZAccessory *)mzAccessory plugin:(CDVPlugin*)plugin{
    self = [super init];
    if (self) {
        self.mzAccessory = mzAccessory;
        self.plugin = plugin;
        self.commandQueue = [[Queue alloc] init];
    }
    return self;
}

// Get autoplay setting
- (void) getAutoPlaySetting:(CDVInvokedUrlCommand*)command {
    if (command == nil){
        return;
    }
    @synchronized
    (self) {
        [self.commandQueue add:command];
    }
    MZAutoPlayCallback callback = ^void(BOOL autoPlay){
        CDVInvokedUrlCommand* callbackCommand = nil;
        @synchronized
        (self) {
            if (![self.commandQueue isEmpty]) {
                callbackCommand = [self.commandQueue poll];
            }
        }
        if (callbackCommand != nil){
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:autoPlay];
            [self.plugin.commandDelegate sendPluginResult:pluginResult callbackId:callbackCommand.callbackId];
        }
    };
    [self.mzAccessory getAutoPlaySettingWithBlock:callback];
}

// Set autoplay setting
- (void) setAutoPlaySetting:(CDVInvokedUrlCommand*)command {
    if (command == nil){
        return;
    }
    @synchronized
    (self) {
        [self.commandQueue add:command];
    }
    BOOL autoPlay = [[command.arguments objectAtIndex:0] boolValue];
    MZAutoPlayCallback callback = ^void(BOOL autoPlay){
        CDVInvokedUrlCommand* callbackCommand = nil;
        @synchronized
        (self) {
            if (![self.commandQueue isEmpty]) {
                callbackCommand = [self.commandQueue poll];
            }
        }
        if (callbackCommand != nil){
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:autoPlay];
            [self.plugin.commandDelegate sendPluginResult:pluginResult callbackId:callbackCommand.callbackId];
        }
    };
    [self.mzAccessory setAutoPlaySetting:autoPlay withBlock:callback];
}

@end